sdks/java/io/rrio/src/main/java/org/apache/beam/io/requestresponse/ThrottleWithExternalResource.java

```diff
@@ -0,0 +1,409 @@
+    return Call.ofCallerAndSetupTeardown(refresherT, VoidCoder.of());
+  }
+
+  private Call<@NonNull T, Void> getEnqueuer() {
+    return Call.ofCallerAndSetupTeardown(enqueuerT, VoidCoder.of());
+  }
+
+  private class ThrottleFn extends DoFn<@NonNull Instant, @NonNull T> {
+    private final String quotaIdentifier;
+    private final DequeuerT dequeuerT;
+    private final DecrementerT decrementerT;
+    private final ReporterT reporterT;
+    private final TupleTag<@NonNull T> outputTag;
+    private final TupleTag<@NonNull ApiIOError> failureTag;
+
+    private ThrottleFn(
+        String quotaIdentifier,
+        DequeuerT dequeuerT,
+        DecrementerT decrementerT,
+        ReporterT reporterT,
+        TupleTag<@NonNull T> outputTag,
+        TupleTag<@NonNull ApiIOError> failureTag) {
+      this.quotaIdentifier = quotaIdentifier;
+      this.dequeuerT = dequeuerT;
+      this.decrementerT = decrementerT;
+      this.reporterT = reporterT;
+      this.outputTag = outputTag;
+      this.failureTag = failureTag;
+    }
+
+    @ProcessElement
+    public void process(@Element @NonNull Instant instant, MultiOutputReceiver receiver) {
+      // Check for available quota.
+      try {
+        if (reporterT.call(quotaIdentifier) <= 0L) {
+          return;
+        }
+
+        // Decrement the quota.
+        Long quotaAfterDecrement = decrementerT.call(instant);
+
+        // As an additional protection we check what the quota is after decrementing. A value
+        // < 0 signals that multiple simultaneous workers have attempted to decrement too quickly.
+        // We don't bother adding the quota back to prevent additional workers from doing the same
+        // and just wait for the next refresh, exiting the DoFn.
+        if (quotaAfterDecrement < 0) {
+          return;
+        }
+
+        // Dequeue an element if quota available. An error here would not result in loss of data
+        // as no element would successfully dequeue from the external resource but instead throw.
+        T element = dequeuerT.call(instant);
+
+        // Finally, emit the element.
+        receiver.get(outputTag).output(element);
+
+      } catch (UserCodeExecutionException e) {
+        receiver
+            .get(failureTag)
+            .output(
+                ApiIOError.builder()
+                    // no request to emit as part of the error.
+                    .setRequestAsJsonString("")
+                    .setMessage(Optional.ofNullable(e.getMessage()).orElse(""))
+                    .setObservedTimestamp(Instant.now())
+                    .setStackTrace(Throwables.getStackTraceAsString(e))
+                    .build());
+      }
+    }
+
+    @Setup
+    public void setup() throws UserCodeExecutionException {
+      enqueuerT.setup();
+      dequeuerT.setup();
+      decrementerT.setup();
+      reporterT.setup();
+    }
+
+    @Teardown
+    public void teardown() throws UserCodeExecutionException {
+      enqueuerT.teardown();
```

Comment:
if there is an exception in previous teardown, should the remaining teardown still get called?

Reply:
Thank you for pointing this out. Could you tell me which of the previous teardown's you are referring?

Reply:
I mean, for example, if `enqueuerT.teardown()` raises exception, then `dequeuerT.teardown()`, `decrementerT.teardown()`, `reporterT.teardown()` won't get called. May consider something like
```
Throwable t = null;
try { enqueuerT.teardown() } catch (...) {t = ...);
try { dequeuerT.teardown() } catch (...) {t = ...);
try { decrementerT.teardown() } catch (...) {t = ...);
try { reporterT.teardown() } catch (...) {t = ...);

if (t != null) throw new UserCodeExecutionException(t);
```

Reply:
Ah now I know what you mean by previously.  Good catch and I agree with you suggestion.